src/nupic/research/boosting.py

```diff
@@ -0,0 +1,270 @@
+    """
+    @param sp - Instance of the Spatial Pooler we want to modify
+    """
+    doUpdateRound = sp._isUpdateRound()
+    doGlobal = sp._globalInhibition or sp._inhibitionRadius > sp._numInputs
+    iteration = sp._iterationNum
+
+    self._updateDutyCycles(overlaps, activeColumns, iteration)
+    # Boosting #1 - boost columns with weak inputs
+    self._bumpUpWeakColumns(sp)
+    self._updateBoostFactors()
+    if doUpdateRound:
+      self._updateMinDutyCycles(sp, doGlobal)
+
+
+  def _updateBoostFactors(self):
+    """
+    Update the boost factors for all columns. The boost factors are used to
+    increase the overlap of inactive columns to improve their chances of
+    becoming active. and hence encourage participation of more columns in the
+    learning process. This is a line defined as: y = mx + b boost =
+    (1-maxBoost)/minDuty * dutyCycle + maxBoost. Intuitively this means
+    that columns that have been active enough have a boost factor of 1, meaning
+    their overlap is not boosted. Columns whose active duty cycle drops too much
+    below that of their neighbors are boosted depending on how infrequently they
+    have been active. The more infrequent, the more they are boosted. The exact
+    boost factor is linearly interpolated between the points (dutyCycle:0,
+    boost:maxFiringBoost) and (dutyCycle:minDuty, boost:1.0).
+
+            boostFactor
+                ^
+    maxBoost _  |
+                |\
+                | \
+          1  _  |  \ _ _ _ _ _ _ _
+                |
+                +--------------------> activeDutyCycle
+                   |
+            minActiveDutyCycle
+    """
+
+    mask = numpy.where(self._minActiveDutyCycles > 0)[0]
+    self._boostFactors[mask] = ((1 - self._maxBoost) /
+      self._minActiveDutyCycles[mask] * self._activeDutyCycles[mask]
+        ).astype(realDType) + self._maxBoost
+
+    self._boostFactors[self._activeDutyCycles > self._minActiveDutyCycles] = 1.0
+
+
+  def _bumpUpWeakColumns(self, sp):
+    """
+    This method increases the permanence values of synapses of columns whose
+    activity level has been too low. Such columns are identified by having an
+    overlap duty cycle that drops too much below those of their peers. The
+    permanence values for such columns are increased.
+
+    @param sp - Instance of the Spatial Pooler we want to modify
+    """
+    weakColumns = numpy.where(self._overlapDutyCycles
+                                < self._minOverlapDutyCycles)[0]
+    for columnIndex in weakColumns:
+      perm = numpy.array([sp.getInputDimensions()], dtype=realDType)
+      sp.getPermanence(columnIndex, perm)
+      pot = numpy.array([sp.getInputDimensions()], dtype=realDType)
+      sp.getPotential(columnIndex, pot)
+      maskPotential = numpy.where(pot > 0)[0]
+      perm[maskPotential] += self._synPermBelowStimulusInc
+      sp._updatePermanencesForColumn(perm, columnIndex, raisePerm=False)
+
+
+# Time keeping methods, counters:
```

Comment:
There is only so much you can define as a convention. I say leave it in for now. If someone thinks it is better to remove later, it will get removed.